### PR TITLE
 arrays-referencing_array_and_its_elements-hard-q2

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [ ] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/arrays/referencing_array_and_its_elements/hard/q2/InThirds.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package arrays.referencing_array_and_its_elements.hard.q2;
+
+/*
+Write a program called InThirds that defines an array of integers, in this case {3, 4, 6, 21, 48, 42, 89, 2, 0}, and
```
Comment (Hedgemon4) [inline src/arrays/referencing_array_and_its_elements/hard/q2/InThirds.java:4]: This question is bad and going to be hard to test thoroughly. Change it to:
`Write a Java program that takes an array of nine numbers from the user and a positive integer x. Then, call a method named remainder(int[] arr, int x) that takes the array and returns the first n numbers in a new array of length n, where n is the remainder after dividing the array length by x. Print the new array in the main method. Assume the user provides valid input. `

File: src/arrays/referencing_array_and_its_elements/hard/q2/InThirds.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+splits the array into three separate subarrays, using a method called splitter that takes and array, split it into
+three equal subarray and return the second subarray.
+For example, if there are 6 elements in the array, then the first two elements go into the first subarray, the middle
+elements go into the second subarray, and the last two elements go into the third subarray. Assume that the number of
+elements in the array is always a multiple of 3. Display all the elements in the second subarray in one line.
+
+ */
+public class InThirds {
```
Comment (Hedgemon4) [inline src/arrays/referencing_array_and_its_elements/hard/q2/InThirds.java:12]: Here is the rewritten test code:
```java
import java.util.Scanner;

public class Remainder {
    public static void main(String[] args) {
        System.out.println("Enter nine integers: ");
        Scanner input = new Scanner(System.in);
        int[] arr = new int[9];
        for (int i = 0; i < arr.length; i++)
            arr[i] = input.nextInt();
        System.out.println("Enter a positive integer: ");
        int x = input.nextInt();
        int[] remain = remainder(arr, x);
        for(int i = 0; i < remain.length; i++)
            System.out.println(remain[i]);
    }

    public static int[] remainder(int[] arr, int x){
        int n = arr.length % x;
        int[] remainder = new int[n];
        for (int i = 0; i < n; i++)
            remainder[i] = arr[i];
        return remainder;
    }
}
```

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                Arguments.of(new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}, answerFor(new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}))
+
+
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("methodInputProvider")
+    void splitterMethodTest(int[] input, int[] ans) throws Throwable {
+        Object output = MethodUtil.invokeIfMethodExists(InThirds.class, "splitter", new Object[]{input}, int[].class);
```
Comment (Hedgemon4) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:84]: Also, switch this to MethodTest

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        int[] a1 = generateRandomArray(-1000, 1000, n);
+        int[] a2 = generateRandomArray(-1000, 1000, n);
+        int[] a3 = generateRandomArray(-1000, 1000, n);
+        return Stream.of(
+                Arguments.of(a1, 3,  Arrays.copyOfRange(a1, 0, n % 3)),
+                Arguments.of(a2, n,  Arrays.copyOfRange(a2, 0, 0)),
+                Arguments.of(a3, 1,  Arrays.copyOfRange(a3, 0, n % 1))
+        );
+    }
+    @ParameterizedTest
```
Comment (Hedgemon4) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:64]: Format this code file (ctrl - alt - l)

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    static Stream<Arguments> InputProvider() {
+        int[] a1 = generateRandomArray(-1000, 1000, n);
+        int[] a2 = generateRandomArray(-1000, 1000, n);
+        int[] a3 = generateRandomArray(-1000, 1000, n);
+        return Stream.of(
+                Arguments.of(a1, 3,  Arrays.copyOfRange(a1, 0, n % 3)),
+                Arguments.of(a2, n,  Arrays.copyOfRange(a2, 0, 0)),
+                Arguments.of(a3, 1,  Arrays.copyOfRange(a3, 0, n % 1))
```
Comment (Hedgemon4) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:61]: You should also a case where n is the divisor, and maybe one more number (like half of n)

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                Arguments.of(a2, n, Arrays.copyOfRange(a2, 0, 0)),
+                Arguments.of(a3, 1, Arrays.copyOfRange(a3, 0, n % 1)),
+                Arguments.of(a4, n, Arrays.copyOfRange(a4, 0, n % n)),
+                Arguments.of(a5, n / 3, Arrays.copyOfRange(a5, 0, n % (n / 3)))
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("InputProvider")
+    void printCorrectOutput(int[] input, int x, int[] ans) throws InvalidClauseException {
```
Comment (Hedgemon4) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:76]:
```suggestion
    void printsCorrectOutput(int[] input, int x, int[] ans) throws InvalidClauseException {
```

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        int[] a1 = generateRandomArray(-1000, 1000, n);
+        int[] a2 = generateRandomArray(-1000, 1000, n);
+        int[] a3 = generateRandomArray(-1000, 1000, n);
+        int[] a4 = generateRandomArray(-1000, 1000, n);
+        int[] a5 = generateRandomArray(-1000, 1000, n);
+        return Stream.of(
+                Arguments.of(a1, 3, Arrays.copyOfRange(a1, 0, n % 3)),
+                Arguments.of(a2, n, Arrays.copyOfRange(a2, 0, 0)),
+                Arguments.of(a3, 1, Arrays.copyOfRange(a3, 0, n % 1)),
+                Arguments.of(a4, n, Arrays.copyOfRange(a4, 0, n % n)),
```
Comment (Hedgemon4) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:69]: I realized this was the same case as one of the previous ones, so maybe change it to n-1 or something instead

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/arrays/referencing_array_and_its_elements/hard/q2/Remainder.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
+package arrays.referencing_array_and_its_elements.hard.q2;
+/*
+Write a Java program that takes an array of nine numbers from the user and a positive integer x. Then, call a method
+named remainder(int[] arr, int x) that takes the array and returns the first n numbers in a new array of length n, where
```
Comment (Hedgemon4) [inline src/arrays/referencing_array_and_its_elements/hard/q2/Remainder.java:4]:
```suggestion
named remainder(int[] arr, int x) that takes the array and returns the first n numbers in a new array of length n, where n
```

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+
+public class MainTest extends BaseTest {
+    // Java
+
+    public static int n = 9;
+
+    public Clause[] testSentence() {
+        TestOption.isInputTest = true;
+        TestOption.defaultInput = "1 2 3 4 5 6 7 8 9 2";
```
Comment (Hedgemon4) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:35]: Generate this instead

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        int[] a2 = generateRandomArray(-1000, 1000, n);
+        int[] a3 = generateRandomArray(-1000, 1000, n);
+        int[] a4 = generateRandomArray(-1000, 1000, n);
+        int[] a5 = generateRandomArray(-1000, 1000, n);
+        int[] a6 = generateRandomArray(-1000, 1000, n);
+        return Stream.of(
+                Arguments.of(a1, 3, Arrays.copyOfRange(a1, 0, n % 3)),
+                Arguments.of(a2, n, Arrays.copyOfRange(a2, 0, 0)),
+                Arguments.of(a3, 1, Arrays.copyOfRange(a3, 0, n % 1)),
+                Arguments.of(a4, n, Arrays.copyOfRange(a4, 0, n % n)),
```
Comment (Hedgemon4) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:70]: Remove this case

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        int[] a2 = generateRandomArray(-1000, 1000, n);
+        int[] a3 = generateRandomArray(-1000, 1000, n);
+        int[] a4 = generateRandomArray(-1000, 1000, n);
+        int[] a5 = generateRandomArray(-1000, 1000, n);
+        return Stream.of(
+                Arguments.of(a1, 3, Arrays.copyOfRange(a1, 0, n % 3)),
+                Arguments.of(a2, n, Arrays.copyOfRange(a2, 0, 0)),
+                Arguments.of(a3, 1, Arrays.copyOfRange(a3, 0, n % 1)),
+                Arguments.of(a4, n - 1, Arrays.copyOfRange(a4, 0, n % (n - 1))),
+                Arguments.of(a5, n - 2, Arrays.copyOfRange(a5, 0, n % (n - 2)))
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:70]:
```suggestion
                Arguments.of(a5, n - 2, Arrays.copyOfRange(a5, 0, 2))
```

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+        };
+
+    }
+
+    public void runMain() {
+        Remainder.main(new String[0]);
+    }
+
+    static Stream<Arguments> InputProvider() {
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:59]: vary range of generated arrays

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+                new Optional(new StringLiteral(" "))
+        };
+
+    }
+
+    public void runMain() {
+        Remainder.main(new String[0]);
+    }
+
+    static Stream<Arguments> InputProvider() {
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:58]:
```suggestion
    static Stream<Arguments> inputProvider() {
```

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+                Arguments.of(a1, 3, Arrays.copyOfRange(a1, 0, n % 3)),
+                Arguments.of(a2, n, Arrays.copyOfRange(a2, 0, 0)),
+                Arguments.of(a3, 1, Arrays.copyOfRange(a3, 0, 0)),
+                Arguments.of(a4, n - 1, Arrays.copyOfRange(a4, 0, 1)),
+                Arguments.of(a5, n - 2, Arrays.copyOfRange(a5, 0, 2))
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("InputProvider")
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:74]:
```suggestion
    @MethodSource("inputProvider")
```

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+        String st = "";
+        for (int i = 0; i < t; i++)
+            st += ans[i] + " ";
+        runWithInput(arrayToInput(input) + " " + x, new Clause[]{
+                new StringLiteral(st.trim())
+        });
+    }
+
+    @ParameterizedTest
+    @MethodSource("InputProvider")
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:86]:
```suggestion
    @MethodSource("inputProvider")
```

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    void printsCorrectOutput(int[] input, int x, int[] ans) throws InvalidClauseException {
+        int t = ans.length;
+        String st = "";
+        for (int i = 0; i < t; i++)
+            st += ans[i] + " ";
+        runWithInput(arrayToInput(input) + " " + x, new Clause[]{
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:80]: There should be an error message set before the runWithInput

File: src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    @MethodSource("InputProvider")
+    void correctRemainderMethod(int[] input, int x, int[] ans) throws Throwable {
+        Object[][] arguments = {
+                {input, int[].class},
+                {x, int.class}
+        };
+        MethodTest m = new MethodTest(Remainder.class, "remainder", arguments);
+        Object output = m.callMethod();
+        CustomAssertions._assertArrayEquals(ans, output,
+                "Your remainder method does not return the correct subarray.");
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/hard/q2/MainTest.java:95]:
```suggestion
                "Your remainder method does not correctly return the subarray of the first n numbers.");
```

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by SamamthaH — APPROVED** [review]

**Pull request merged**